website/blog/2024-08-14-otel-context-propagation-in-ferretdb.md

```diff
@@ -0,0 +1,208 @@
+---
+slug: otel-context-propagation-in-ferretdb
+title: 'OpenTelemetry context propagation in FerretDB'
+authors: [elena]
+description: >
+  In this blog post, we demonstrate how to pass tracing context to queries in FerretDB using OpenTelemetry.
+tags: [observability, opentelemetry]
+---
+
+In today's world of distributed systems, achieving reliability depends on various factors, one of which is effective observability.
+OpenTelemetry (OTel) has emerged as a standard for distributed tracing, but passing context to databases remains a significant challenge,
+particularly with document databases like MongoDB.
+At FerretDB, we're committed to addressing this challenge.
+
+<!--truncate-->
+
+While adopting OpenTelemetry in application development is relatively straightforward across various programming languages,
+the process becomes more complex when passing context to databases.
+Most databases don't natively support tracing-related context.
+Approaches like [SQLCommenter](https://google.github.io/sqlcommenter/) have been developed to bridge this gap,
+enabling the connection between current trace data and database queries.
+In these cases, trace information is injected at the ORM level and passed to the database via SQL comments.
+Database operators can then enable query logs to link specific queries with the corresponding application requests.
+
+But is it possible to achieve something similar with MongoDB's query language?
+At FerretDB, we believe that enabling the linking of a query's lifecycle with the caller's response can help developers make
+their applications more predictable and reliable.
+
+Let's consider a SQLCommenter-like approach for MongoDB.
+We can leverage MongoDB's existing features to attach metadata to queries.
+One such feature is the comment field, which allows additional information to be included in MongoDB queries.
+This field doesn't influence query logic but can be used to provide valuable context.
+
+At FerretDB, we want to empower users to pass such context.
+We decided to utilize the `comment` field to do it.
+We parse the content of the `comment` field, and if it's a JSON document with a `ferretDB` key, we check for tracing data.
+If the tracing data is present, FerretDB sets the parent's context when a span for the current operation is created.
+
+An example of a comment with tracing data could look like this:
+
+```json
+{
+  "ferretDB": {
+    "traceID": "1234567890abcdef1234567890abcdef",
+    "spanID": "fedcba9876543210"
+  }
+}
+```
+
+The strings `traceID` and `spanID` are hex-encoded strings that represent the corresponding trace and span identifiers.
+For example, the JSON above represents the following data:
+
+```go
+traceID := [16]byte{0x12, 0x34, 0x56, 0x78, 0x90, 0xab, 0xcd, 0xef, 0x12, 0x34, 0x56, 0x78, 0x90, 0xab, 0xcd, 0xef}
+spanID := [8]byte{0xfe, 0xdc, 0xba, 0x98, 0x76, 0x54, 0x32, 0x10}
+```
```

Comment:
I think that representation makes thing more confusing coming after JSON above, not more clear

Reply:
removed